deploy.go

```diff
@@ -0,0 +1,200 @@
+package goscaleio
+
+import (
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"io"
+	"mime/multipart"
+	"net/http"
+	"os"
+	path "path/filepath"
+)
+
+var (
+	errNewClient = errors.New("missing endpoint")
+	errSysCerts  = errors.New("Unable to initialize cert pool from system")
+)
+
+// GatewayClient is client for Gateway server
+type GatewayClient struct {
+	http     *http.Client
+	host     string
+	username string
+	password string
+}
+
+// NewGateway returns a new gateway client.
+func NewGateway(
+	host string, username, password string, insecure, useCerts bool) (*GatewayClient, error) {
+
+	if host == "" {
+		return nil, errNewClient
+	}
+
+	gc := &GatewayClient{
+		http:     &http.Client{},
+		host:     host,
+		username: username,
+		password: password,
+	}
+
+	if insecure {
+		gc.http.Transport = &http.Transport{
+			/* #nosec G402 */
+			TLSClientConfig: &tls.Config{
+				InsecureSkipVerify: true,
+			},
+		}
+	}
+
+	if !insecure || useCerts {
+		pool, err := x509.SystemCertPool()
+		if err != nil {
+			return nil, errSysCerts
+		}
+
+		gc.http.Transport = &http.Transport{
+			/* #nosec G402 */
+			TLSClientConfig: &tls.Config{
+				RootCAs:            pool,
+				InsecureSkipVerify: insecure,
+			},
+		}
+	}
+
+	return gc, nil
+}
+
+// GatewayFunction is an  interface which has all the functionalities for the gateway.
+type GatewayFunction interface {
+	UploadPackages(fliePath string) error
+	ParseCSV(filePath string) error
+	BeginInstallation(jsonStr, mdmUsername, mdmPassword, liaPassword string) error
+}
+
+// UploadPackages used for upload packge to gateway server
+func (gc *GatewayClient) UploadPackages(filePath string) error {
+
+	file, err1 := os.Open(path.Clean(filePath))
+	if err1 != nil {
+		return err1
+	}
+	defer func() error {
+		if err := file.Close(); err != nil {
+			return err
+		}
+		return nil
+	}()
+	body := &bytes.Buffer{}
+	writer := multipart.NewWriter(body)
+
+	part, err2 := writer.CreateFormFile("files", path.Base(filePath))
+	if err2 != nil {
+		return err2
+	}
+	_, err3 := io.Copy(part, file)
+	if err3 != nil {
+		return err3
+	}
+	err4 := writer.Close()
+	if err4 != nil {
+		return err4
+	}
+
+	req, err5 := http.NewRequest("POST", gc.host+"/im/types/installationPackages/instances/actions/uploadPackages", body)
+	if err5 != nil {
+		return err5
+	}
+	req.Header.Set("Content-Type", writer.FormDataContentType())
+	req.Header.Set("Authorization", "Basic "+base64.StdEncoding.EncodeToString([]byte(gc.username+":"+gc.password)))
+	client := gc.http
+	_, err6 := client.Do(req)
+	if err6 != nil {
```

Comment:
can you please rename these err1,err2,... err6 to something meaningful error names.

Reply:
Updated.